plugins/in_winevtlog/winevtlog.c

```diff
@@ -0,0 +1,786 @@
+    if (size == 0)
+        return NULL;
+
+    buf = flb_malloc(size);
+    if (buf == NULL) {
+        flb_errno();
+        return NULL;
+    }
+    size = WideCharToMultiByte(codePage, 0, wstr, -1, buf, size, NULL, NULL);
+    if (size == 0) {
+        flb_free(buf);
+        return NULL;
+    }
+
+    return buf;
+}
+
+/*
+ * Load the bookmark from SQLite DB.
+ */
+int winevtlog_sqlite_load(struct winevtlog_channel *ch, struct flb_sqldb *db)
+{
+    int ret;
+    char query[1024];
+    struct winevtlog_sqlite_record record = {0};
+    EVT_HANDLE hBookmark = NULL;
+    PWSTR wBookmarkXML = NULL;
+    struct winevtlog_channel *re_ch = NULL;
+
+    snprintf(query, sizeof(query) - 1, SQL_GET_CHANNEL, ch->name);
+
+    ret = flb_sqldb_query(db, query, winevtlog_sqlite_callback, &record);
+    if (ret == FLB_ERROR) {
+        return -1;
+    }
+
+    if (record.created) {
+        ch->time_created = record.created;
+    }
+    if (record.time_updated) {
+        ch->time_updated = record.time_updated;
+    }
+
+    if (record.name) {
+        wBookmarkXML = convert_str(record.bookmark_xml);
+        if (wBookmarkXML) {
+            hBookmark = EvtCreateBookmark(wBookmarkXML);
+            if (hBookmark) {
+                /* re-create subscription handles */
+                re_ch = winevtlog_subscribe(ch->name, FLB_FALSE, hBookmark);
+                if (re_ch != NULL) {
+                    close_handles(ch);
+
+                    ch->bookmark = re_ch->bookmark;
+                    ch->subscription = re_ch->subscription;
+                    ch->signal_event = re_ch->signal_event;
+                }
+                else {
+                    flb_error("Failed to subscribe with bookmarkXML: %s\n", record.bookmark_xml);
+                    ch->bookmark = EvtCreateBookmark(NULL);
+                }
+            }
+            else {
+                flb_error("Failed to load bookmarkXML with %d\n", GetLastError());
+                ch->bookmark = EvtCreateBookmark(NULL);
+            }
+        }
+        if (wBookmarkXML) {
+            flb_free(wBookmarkXML);
+        }
+    }
+    return 0;
+}
+
+/*
+ * Save the bookmark into SQLite DB.
+ */
+int winevtlog_sqlite_save(struct winevtlog_channel *ch, struct flb_sqldb *db)
+{
+    int ret;
+    char query[1024];
+    PWSTR wBookmarkXML = NULL;
+    char *bookmarkXML;
+    int used_size = 0;
+
+    wBookmarkXML = render_event(ch->bookmark, EvtRenderBookmark, &used_size);
+    if (wBookmarkXML == NULL) {
+        flb_error("failed to render bookmark with %d", GetLastError());
+        flb_free(wBookmarkXML);
+
+        return -1;
+    }
+    bookmarkXML = convert_wstr(wBookmarkXML, CP_UTF8);
+    if (bookmarkXML == NULL) {
+        flb_error("failed to convert Wider string with %d", GetLastError());
+        flb_free(wBookmarkXML);
+        flb_free(bookmarkXML);
+
+        return -1;
+    }
+
+    snprintf(query, sizeof(query) - 1, SQL_UPDATE_CHANNEL,
```

Comment:
Should we escape these strings to prevent uninentional sql injection from breaking the query and causing something to be missed? Is that a possibility with the type of data that rnder_event generates?

Reply:
XML encoded channel bookmark information is generated by Windows Event Tracing system.
Also, I'm not sure what should we do to protect suspicious XML encoded bookmarks....

Reply:
That's why I mentioned it unintentionally breaking quotation marks, because the character might just be a legitimate part of the data returned.

Reply:
The actual stored channel bookmarks and metadata in the sqlite DB as follows:

```console
sqlite> select * from in_winevtlog_channels;
Application|<BookmarkList>
  <Bookmark Channel='Application' RecordId='33640' IsCurrent='true'/>
</BookmarkList>|1637333345|1633878463
Security|<BookmarkList>
  <Bookmark Channel='Security' RecordId='497867' IsCurrent='true'/>
</BookmarkList>|1637333345|1633878464
System|<BookmarkList>
  <Bookmark Channel='System' RecordId='32359' IsCurrent='true'/>
</BookmarkList>|1637333349|1633878464
Microsoft-Windows-Time-Service/Operational|<BookmarkList>
  <Bookmark Channel='Microsoft-Windows-Time-Service/Operational' RecordId='1800' IsCurrent='true'/>
</BookmarkList>|1637297649|1633878464
Microsoft-Windows-SMBClient/Connectivity|<BookmarkList>
  <Bookmark Channel='Microsoft-Windows-SmbClient/Connectivity' RecordId='23396' IsCurrent='true'/>
</BookmarkList>|1637333345|1633878465
```

The following structure is fixed:

```xml
<BookmarkList>
  <Bookmark Channel='[CHANNEL NAME]' RecordId='[RECORD NUMBER]' IsCurrent='true'/>
</BookmarkList>
```

Reply:
And that works because it's predictable data and because of that for that specific field you used double quotation marks as delimiter which is totally fine.